wgpu-core/src/lock/observing.rs

```diff
@@ -0,0 +1,330 @@
+//! Lock types that observe lock acquisition order.
+//!
+//! This module's [`Mutex`] type is instrumented to observe the
+//! nesting of `wgpu-core` lock acquisitions. Whenever `wgpu-core`
+//! acquires one lock while it is already holding another, we note
+//! that nesting pair. This tells us what the [`LockRank::followers`]
+//! set for each lock would need to include to accommodate
+//! `wgpu-core`'s observed behavior.
+//!
+//! When `wgpu-core`'s `observe_locks` feature is enabled, if the
+//! `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is set to the
+//! path of an existing directory, then every thread that acquires a
+//! lock in `wgpu-core` will write its own log file to that directory.
+//! You can then run the `wgpu` workspace's `analyze-locks` binary
+//! to read those files and summarize the results. The output from
+//! `analyze-locks` has the same form as the lock ranks given in
+//! [`lock/rank.rs`].
```

Comment:
I wanted to link to the source code, since that's the clearest presentation of the ranks.

Reply:
Just validated that this works locally, so no blocking concern here.

I think it would be more direct to link to `LockRankNumber`, and that the entry for `LockRankNumber` suggest examining the `source` link directly:

![image](https://github.com/gfx-rs/wgpu/assets/658538/f78b21c1-6655-4146-953a-a8bf82fee5df)

…so that the correct line numbers are highlighted, and the user doesn't need to hunt around:

![image](https://github.com/gfx-rs/wgpu/assets/658538/ed4d954f-92f6-4aa5-8127-3abac3f2c4b8)

…but I'm happy to iterate on that in a follow-up PR, if you're interested.